### ISSUE TITLE
[performance] nnvm.compiler.build hasn't brought enough improvements

### ISSUE THREAD

**Issue by BigJerry** [dmlc/nnvm#250]
Below is my network's structure (input shape is (1,1,512,640)), mainly composing of separable depthwise convolution layer BN and relu, which is very similar to mobilenet. however , somehow after nnvm.compiler.build being called ,time consumption is 130 ms, whereas before nnvm.compiler.build 's optimization , the time needed for inference is 140ms. by the way , my GPU is 1050 GTX.. here: http://tvmlang.org/2017/08/22/Optimize-Deep-Learning-GPU-Operators-with-TVM-A-Depthwise-Convolution-Example.html , in this blog the improvement for depthwise convolution is remarkable. can nnvm.compiler.build method achieve similar result ?

def conv_block(data, name, depthwise_channels, pointwise_channels, layout, kernel_size=(3,3), strides=(1,1), padding=[1,1]):
    """This building block is composed of separable convolutional layer and activation layer."""
    depthconv = sym.conv2d(data=data, channels=depthwise_channels,groups=depthwise_channels,
                      kernel_size=kernel_size,strides=strides,
                      padding=padding,use_bias=False,layout=layout,name=name+"1_depthwise_conv")
    pointconv = sym.conv2d(data=depthconv, channels=pointwise_channels,
                           kernel_size=(1,1), strides=strides,
                           padding=[0,0], use_bias=True, layout=layout, name=name+"1_pointwise_conv")
    relu = sym.relu(data = pointconv, name=name+"_relu")
    return relu

def separable_conv_block(data, name, depthwise_channels, pointwise_channels, layout, strides = (1,1),kernel_size=(3,3), padding=[1,1],epsilon=1e-5):
    """This building block is composed of separable convolutional layer, batch normalization layer and activation layer."""
    depthconv = sym.conv2d(data=data, channels=depthwise_channels,groups=depthwise_channels,
                      kernel_size=kernel_size,strides=strides,padding=padding,use_bias=False,
                      layout=layout,name=name+"_depthwise_conv")
    pointconv = sym.conv2d(data=depthconv, channels=pointwise_channels, kernel_size=(1,1),
                           strides=strides, padding=[0,0],use_bias=True,
                            layout=layout, name=name+"_pointwise_conv")
    bn = sym.batch_norm(data=pointconv,epsilon=epsilon,name=name+"_batchnorm",center=True,scale=True,axis=1)
    relu = sym.relu(data=bn, name=name+"_relu")
    return relu
    
def dncnn_net(num_separable_conv_block=5,layout='nchw'):
    """Defining what the topology is for CNN."""
    data = sym.Variable("data")
    body = conv_block(data, "separable_conv_block_", 1,64,layout)
    for n in range(num_separable_conv_block):
        body = separable_conv_block(body, "separable_conv_block_"+str(n+2), 64,64,layout)
    body = sym.conv2d(data=body, channels=1, groups=1, kernel_size=(3,3),strides=(1,1),
                      padding=[1,1],use_bias=True,layout=layout, name="last_conv")
    return body

**Comment by Huyuwei** [comment]
That blog is operator level optimization. nnvm.compiler.build's optimization is graph level, and you can set the opt_level: https://github.com/dmlc/nnvm/blob/master/python/nnvm/compiler/build_module.py#L11

In your case, the acceleration from 140ms to 130ms is brought by graph level optimization.

**Comment by BigJerry** [comment]
@Huyuwei do you mean operator level optimization has to be done on user's side ,following tvm's tutorial or the instructions in your blog?

**Comment by Huyuwei** [comment]
@BigJerry  nnvm compiler calls into computing kernels in [TOPI](https://github.com/dmlc/tvm/tree/master/topi). The operators in TOPI are already optimized and expected to be reused. Of course, you can further optimize it, tune it for specific workloads, or add new operators.

**Comment by BigJerry** [comment]
@Huyuwei thanks a lot! I'll try it .

**Comment by BigJerry** [comment]
@Huyuwei still have a question for you . I used nnvm.compiler.build method to obtain graph level optimization. Did you mean that if I want operator level optimization , I have to struct my network via tvm.topi instead of nnvm.symbol ?

**Comment by Huyuwei** [comment]
nnvm.compiler.build performs graph optimization and then calles into TOPI to do the real computing. They work together.

I assume that by saying operator level optimization, you mean improving the performance of a single operator, e.g. depthwise conv and conv. This should be done in TOPI, by modifying the operators' schedules.

**Comment by Huyuwei** [comment]
@BigJerry You should build the network using nnvm.symbol, since it's on the graph symbol level. It will call into topi to do the real computing part.

**Comment by BigJerry** [comment]
@Huyuwei as you can see , I've already constructed a network  with building blocks composing of  depthwise convolution , but it didn't perform well like the depthwise convolution in your blog . What should I do to reach the remarkable speed you'd achieved as my network already has been built via nnvm.symbol system?

**Comment by Huyuwei** [comment]
The net is composed of depthwise conv and also 1x1 conv. Since conv is much more time consuming than depthwise conv, it's probable that the bottleneck is 1x1 conv. You can use nvprof to do a profiling and measure the time cost of each operator kernel.

**Comment by BigJerry** [comment]
Got it . I'll do some tests. by the way , we share same age(from your home page). you and TianQi Chen are my models. @Huyuwei

**Comment by BigJerry** [comment]
here is my WeChat Id ZZW971749411 , it will be very appreciated if you could add me into your list. @Huyuwei

**Comment by Huyuwei** [comment]
haha, Ok @BigJerry
